mpqp/execution/providers/google.py

```diff
@@ -0,0 +1,340 @@
+from __future__ import annotations
+from typing import Optional
+
+from typeguard import typechecked
+
+from mpqp.execution.devices import GOOGLEDevice
+from mpqp.execution.job import JobType, Job
+from mpqp.execution.result import Result, Sample, StateVector
+
+from mpqp.core.instruction.measurement import ComputationalBasis
+from mpqp.core.instruction.measurement.basis_measure import BasisMeasure
+from mpqp.core.instruction.measurement.expectation_value import ExpectationMeasure
+
+from mpqp import Language
+
+from cirq.work.observable_measurement_data import ObservableMeasuredResult
+from cirq.sim.state_vector_simulator import StateVectorTrialResult
+from cirq.transformers.optimize_for_target_gateset import optimize_for_target_gateset
+from cirq.value.probability import state_vector_to_probabilities
+from cirq.sim.sparse_simulator import Simulator
+from cirq.circuits.circuit import Circuit as Cirq_circuit
+from cirq.study.result import Result as cirq_result
+from cirq.ops.linear_combinations import PauliSum as Cirq_PauliSum
+from cirq_google.engine.virtual_engine_factory import (
+    load_median_device_calibration,
+    create_device_from_processor_id,
+)
+from cirq_google.engine.simulated_local_processor import SimulatedLocalProcessor
+from cirq_google.engine.simulated_local_engine import SimulatedLocalEngine
+from qsimcirq.qsim_simulator import QSimSimulator
+from cirq.work.observable_measurement import (
+    measure_observables,
+    RepetitionsStoppingCriteria,
+)
+import cirq_ionq as ionq
+from cirq_ionq.ionq_gateset import IonQTargetGateset
+from cirq.devices.line_qubit import LineQubit
+
+
+@typechecked
+def run_google(job: Job) -> Result:
+    """
+    Execute the job on the right Google device precised in the job in parameter.
+    This function is not meant to be used directly, please use ``runner.run(...)`` instead.
+
+    Args:
+        job: Job to be executed.
+
+    Returns:
+        A Result after submission and execution of the job.
+    """
+    return run_local(job) if not job.device.is_remote() else run_google_remote(job)
+
+
+@typechecked
+def run_google_remote(job: Job) -> Result:
+
+    assert type(job.device) == GOOGLEDevice
+
+    job_cirq_circuit = job.circuit.to_other_language(Language.CIRQ)
+    assert isinstance(job_cirq_circuit, Cirq_circuit)
+
+    if job.device.is_ionq():
+        if job.job_type != JobType.SAMPLE:
+            raise ValueError(
+                f"{job.device}: job_type must be {JobType.SAMPLE} for job type {job.job_type}"
```

Comment:
Maybe to be clearer we can say
`f"{job.device}: job_type must be {JobType.SAMPLE} but go job type {job.job_type}"`